File: TheHungryJava/app/src/main/java/xyz/imkaem/thehungryjava/MainActivity.java
package xyz.imkaem.thehungryjava;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    CardView cardStarters;
    CardView cardMains;
    CardView cardDesserts;
    TextView textViewSendEmail;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

//        assign view variables
        cardStarters = findViewById(R.id.card_view_starters);
        cardMains = findViewById(R.id.card_view_main_courses);
        cardDesserts = findViewById(R.id.card_view_desserts);
        textViewSendEmail = findViewById(R.id.text_view_send_email);

        cardStarters.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent startersActivityIntent = new Intent(MainActivity.this, StartersActivity.class);
                startActivity(startersActivityIntent);

            }
        });

        cardMains.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // now we create the intent - pass it context, and where to go
                Intent mainsActivityIntent = new Intent(MainActivity.this, MainCoursesActivity.class);

                startActivity(mainsActivityIntent);

                // and then we need to call to start activity
            }
        });

        cardDesserts.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // now we create the intent - pass it context, and where to go
                Intent mainsActivityIntent = new Intent(MainActivity.this, DessertsActivity.class);

                startActivity(mainsActivityIntent);

                // and then we need to call to start activity
            }
        });

        textViewSendEmail.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                // now we need to create implicit intent
                Intent sendEmailIntent = new Intent(
                        Intent.ACTION_SENDTO
                );
                // with this, only email apps should be handling this intent
                sendEmailIntent.setData(Uri.parse("mailto:theinfo@hungryjava.com"));

                startActivity(sendEmailIntent);
            }
        });
    }
}

File: lifecycle_app/app/src/main/java/xyz/imkaem/lifecycleapp/MainActivity.kt
package xyz.imkaem.lifecycleapp

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.os.PersistableBundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.activity.addCallback
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import xyz.imkaem.lifecycleapp.databinding.ActivityMainBinding
import java.io.File
import java.util.*

class MainActivity : AppCompatActivity() {

    private lateinit var viewBiding: ActivityMainBinding
    private var isFirstLoad: Boolean = true

    var seconds = 0;
    private lateinit var refreshTimer: Timer

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

//        prevent going back - oir intercepting it
        val callback = onBackPressedDispatcher.addCallback {

            Toast.makeText(this@MainActivity, "Hello", Toast.LENGTH_LONG).show()
            showDialog()

        }

        viewBiding = ActivityMainBinding.inflate(layoutInflater)
        val view = viewBiding.root
        setContentView(view)

        viewBiding.buttonToAnotherActivity.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                val intent: Intent = Intent(this@MainActivity, AnotherActivity::class.java)
                startActivity(intent)
            }

        })

        viewBiding.buttonExitActivity.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
//                finish()

                showDialog()
            }
        })

//        save text to text
        viewBiding.buttonSaveMessage.setOnClickListener {
            saveMessage()
        }

//        we can set it here because we will set updated ones in onRestart
        viewBiding.textViewRefreshStatus.text = "Hello - status is 'onCreate'"

//        setContentView(R.layout.activity_main)

        Log.d("karlo", "onCreate() called")

//        TODO this is some timer
//        val timer: CountDownTimer = object : CountDownTimer(2000, 100) {
//            override fun onTick(millisUntilFinished: Long) {
//                println("Ticking")
//            }
//
//            override fun onFinish() {
//                Log.d("karlo", "onFinish() called")
//                finish()
//            }
//        }

//        timer.start()

//        this is another timer
//        var seconds = 0;
//        fixedRateTimer(period = 1000L) {
////            we need this to be allowed to touch UI
//            runOnUiThread {
//                seconds++
//                viewBiding.textViewLengthStatus.text = "More text: $seconds"
//            }
//        }

//        THIS IS IN onCreate
        val textViewSavedMessage: String = savedInstanceState?.getString("savedTextViewMessage") ?: ""
        viewBiding.textViewMessage.text = textViewSavedMessage
    }

    private fun saveMessage() {
        val userMessage = viewBiding.editTextMessage.text.toString()
        viewBiding.textViewMessage.text = userMessage
    }

    override fun onStart() {
        super.onStart()
        Log.d("karlo", "onStart() called")
    }

    override fun onResume() {
        super.onResume()
        Log.d("karlo", "onResume() called")
//  this is better - do it in onresume
//        val data: String = when (isFirstLoad) {
//            true -> "This was initial data"
//            false -> "This was data set when app resumed"
//        }
////        TODO this is where we want to refresh
//        viewBiding.textViewRefreshStatus.text = data
//        isFirstLoad = false

//        refreshTimer = fixedRateTimer(period = 1000L) {
////            we need this to be allowed to touch UI
//            runOnUiThread {
//                seconds++
//                viewBiding.textViewLengthStatus.text = "More text: $seconds"
//            }
//        }
    }

    override fun onRestart() {
        super.onRestart()

        // not that common - better to do it in onResume
        viewBiding.textViewRefreshStatus.text = "Hello - status is 'onRestart'"
    }

    override fun onPause() {
        super.onPause()

        Log.d("karlo", "onPause() called")

//        refreshTimer.cancel()
    }

    override fun onStop() {
        super.onStop()
        Log.d("karlo", "onStop() called")
    }

//    override fun onDestroy() {
//        Log.d("karlo", "onDestroy() called")
//        val message: String = viewBiding.editTextMessage.text.toString()
//
//        val file: File = File(filesDir, "message.txt")
//        file.writeText(message)
//
//        super.onDestroy()
//    }

//    override fun onBackPressed() {
////        this makes app not leave the app, if removed
////        super.onBackPressed()
//    }

    fun showDialog() {
        AlertDialog.Builder(this).setTitle("Some Dialog").setMessage("Some message")
            .setView(R.layout.dialog_warning)
            .setPositiveButton(
                "Yes"
            ) { dialog, which -> finish() }
            .setNegativeButton("No", object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface?, which: Int) {
                    Toast.makeText(this@MainActivity, "No clicked", Toast.LENGTH_LONG).show()
                    dialog?.dismiss()
                }
            })
            .setNeutralButton("Help", object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface?, which: Int) {
                    Toast.makeText(this@MainActivity, "Help clicked", Toast.LENGTH_LONG).show()
                    dialog?.dismiss()
                }

            })
            .show()

    }

    //    onsaveinstancestate
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

//
        val savedTextViewMessage: String = viewBiding.textViewMessage.text.toString()
        outState.putString("savedTextViewMessage", savedTextViewMessage)

    }
}